[mta] APEX style Fused Adam

crcrpar opened the pull request:
This PR implements an APEX style FusedAdam in PyTorch.
This is different from the APEX one in that this is compatible with `torch.cuda.amp.GradScaler` by setting `_step_supports_amp_scaling` to `True` and unscales gradients inside its CUDA kernel.

related: https://github.com/pytorch/pytorch/issues/68041, https://github.com/pytorch/pytorch/issues/71274, https://github.com/pytorch/pytorch/issues/80167
possibly related to https://github.com/pytorch/pytorch/issues/80595#issuecomment-1178519436

cc @ptrblck @ngimel

crcrpar committed:
Add signature of `_fused_adam_` to at::native namespace

In order to conform with the codegen logic, `params` are written as
`self` in `native_functions.yaml`.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Scalar -> float

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Draft: Real FusedAdamFunctor

Build is failing though.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
still failing

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Fix build

- `init_args` did not get update to support
`FusedOptimizerTensorListMetadata`
- I mistakenly tell the functor that metadata is
`FusedOptimizerTensorListMetadata`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
init FusedAdam class

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Support `maximize` and `amsgrad`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
`_fused_adam` -> `_fused_adam_` & misc

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Add initial test cases

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
correctly iterate over param_group['params']

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Misc fix

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Allow for `capturable=True`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Add cases of FusedAdam

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cosmetic fix: use alias `optim._fused` -> `optim_fused`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Add GradScaler & FusedAdam cases

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
fix typo

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
missing import

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remove fusedAdam as it doesn't support CPU path

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
set `_step_supports_amp_scaling=True`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
add case of graph, fusedadam, gradscaler

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
make const as possible

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
make `step` scalar Tensor

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Support both path after extracting a math into a device function

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
fix the value of `autogen`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
pass state_steps as TensorList

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
fix logic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
`state_steps` better be 1D for fused adam

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Check the behavior when `found_inf` is a tensor of ones.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cuda graph tests

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
fix cases

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
[note] feature of having higher prec copy of params

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
update error message

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
TODO: group tensors by device & dtype likewise GradScaler

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
support params on different devices in one params_group

following
https://github.com/pytorch/pytorch/blob/36d2c44cce027be82dca6fd8134736ec47e1acca/torch/cuda/amp/grad_scaler.py#L198-L226

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
duplicate grads to avoid inplace updates in graph capture cases

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cosmetic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cosmetic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
write gradients iff inv_grad_scale

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
revert duplication of grads

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
calculate inverse of grad_scaler in CUDA kernel

This will merge `at.empty` and fill into optimizer.step

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
adjustable atol in grad_scaling_autocast

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
update includes

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Avoid `rand` to make test a bit faster

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
TestCuda::test_grad_scaling_autocast_fusedadam is failing

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
guarantee the parity with native Adam under AMP

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Stop passing redundant `capturable` to `_fused_adam_`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Reimplement MultiDeviceReplicator for Torch Script

`torch.device` is not compatible with torch script.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remove conditional includes

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

ngimel commented:
Hey it's great that you've added special tests for cuda graphs, given that we previously had issues with adam + cudagraphs, can you please describe here how you are dealing with the changing cpu state.

crcrpar commented:
> given that we previously had issues with adam + cudagraphs, can you please describe here how you are dealing with the changing cpu state.

It was straightforward except having `step` tensors on CUDA following https://github.com/pytorch/pytorch/pull/77862. I guess what was complicating was around `bias_correction1` and `bias_correction2` like https://github.com/pytorch/pytorch/blob/master/torch/optim/adamw.py#L354-L356 which this kernel does in CUDA kernel.

crcrpar committed:
Prefix `_` to MultiDeviceReplicator as it doesn't need to be public.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

ngimel commented:
Hey @crcrpar, there's the issue with updating max values in mta adam #78807 - does this PR handle the update correctly? Can we make sure there are tests for that? If this PR fixes it, we could swap mta implementation to use this PR to fix the issue.

crcrpar commented:
I think I checked parity of states while implementing this but let me double check.
and excuse me for delaying the fix for msvc build.

(related to #81894)

ngimel commented on aten/src/ATen/native/cuda/FusedAdamKernel.cu:
```diff
@@ -0,0 +1,228 @@
+        opmath_t param = static_cast<opmath_t>(r_args[kParamIdx][ii]);
+        opmath_t grad = static_cast<opmath_t>(r_args[kGradIdx][ii]);
+        if (grad_scale_ptr) {
+            grad /= (static_cast<double>(*grad_scale_ptr));
```
do you need a double here? Also, is grad_scale_ptr exposed in public API? Please leave a comment that it's unused if it isn't. 
Edit: oh I see, some versions do support grad scaling

crcrpar replied:
I'm not sure but `GradScaler.unscale_` casts its scale to double when taking reciprocal:

https://github.com/pytorch/pytorch/blob/2f9d046d6717d9423e012c5249624fbeeb506cbb/torch/cuda/amp/grad_scaler.py#L277

ngimel commented on aten/src/ATen/native/cuda/FusedAdamKernel.cu:
```diff
@@ -0,0 +1,228 @@
+        if (weight_decay != 0) {
+            grad += param * weight_decay;
+        }
+        exp_avg = beta1 * exp_avg + (1 - beta1) * grad;
```
nit: you could use fma version here https://developer.nvidia.com/blog/lerp-faster-cuda/, it's slightly more accurate.

ngimel commented on aten/src/ATen/native/cuda/FusedAdamKernel.cu:
```diff
@@ -0,0 +1,228 @@
+// On the other hand, fused optimizers set their member variable of `_step_supports_amp_scaling` to `True`
+// in order to remove the device sync above. This means that fused optimizers have to have
+// their CUDA kernels (a) unscale gradients and (b) skip parameter updates accordingly.
+// To be functionally on par with `torch.optim` optimizers and `_multi_tensor` ones,
```
this is confusing, because currently the interface to optimizers is constructing via torch.optim and setting a flag whether to use mta. In this sense, regular and mta optimizers are functionally equivalent (and mta optimizers typically don't support additional functionality like scaling)

crcrpar replied:
um sorry I don't understand. probably due to my bad comment.
What I meant to say is, when `grad_scaler.step(optimizer)` is called and this optimizer is Adam or foreach Adam, the grad scaler sees and unscales the gradient values to call `optimizer.step` (= _single_adam or _multi_tensor_adam function).
On the other hand, if it's fused Adam which sets `_step_supports_amp_scaling` to `True`, `optimizer.step` is called notwithstanding gradients' sanity, so _fused_adam (the functional form) and the cuda kernel need to skip the param update itself and take care of step counts.

ngimel commented on test/test_cuda.py:
```diff
@@ -2322,10 +2325,42 @@ def run(data, model, optimizer, scaler, loss_fn, skip_iter, try_scaling_api):
+            self.assertEqual(loss_control, loss_scaling)
+            for param_control, param_scaling in zip(mod_control.parameters(), mod_scaling.parameters()):
+                self.assertEqual(param_control.grad, param_scaling.grad)
+                self.assertEqual(param_control, param_scaling)
```
should you also check exp_avg buffers here?

ngimel commented on torch/optim/adam.py:
```diff
@@ -405,3 +487,85 @@ def _multi_tensor_adam(params: List[Tensor],
+# TODO(crcrpar): Think of rewriting this in C++.
+@torch.no_grad()
+def _group_params_by_device_and_dtype(
+    params: List[Tensor],
```
you probably can make this function more generic by using *args instead of named arguments, and just iterating over arguments list inside.

ngimel commented on aten/src/ATen/native/cuda/FusedAdamKernel.cu:
```diff
@@ -0,0 +1,228 @@
+
+        // Store results.
+        r_args[kParamIdx][ii] = param;
+        r_args[kGradIdx][ii] = grad_to_store;
```
why are you unconditionally storing grad even if it wasn't scaled?

crcrpar replied:
good catch. No reasons, I forgot to update here.

ngimel commented on torch/optim/_fused/__init__.py:
```diff
@@ -0,0 +1,18 @@
+"""
+:mod:`torch.optim._fused` is a package implementing various optimization algorithms.
```
why do we need new _fused class instead of a `fused=True` arg to regular optimizer (which is how it's implemented anyway?)

crcrpar replied:
I mimicked `torch.optim._multi_tensor` ones. if you don't think we need this, I'll remove this `_fused` directory.

ngimel replied:
`_multi_tensor` are no longer in use, they are accessible through regular ones with a constructor arg. Also, please update docstring for adam/adamw optimizers to describe the new flag.

crcrpar replied:
Understood, so it looks like they are there mainly for test_optim.py at the moment.

ngimel replied:
`test_optim` should also be rewritten to use regular optimizers with `multi_tensor` kwarg, and not `_multi_tensor`.

crcrpar replied:
there's still some `optim._multi_tensor` uses in test_optim.py but I accessed this to some extent in https://github.com/pytorch/pytorch/pull/81705/commits/97c106a69c1f5dcb33aff33982b95e12ccbd89bb and removed `torch.optim._fused` in 87dc7b32fbccd6b84691a54704355d7c3ef95d07.

crcrpar replied:
https://github.com/pytorch/pytorch/pull/83549 can be a follow-up

ngimel reviewed (commented):
This generally looks good, but there are conflicts now, and still some build errors?

crcrpar commented on torch/optim/adam.py:
```diff
@@ -405,3 +487,85 @@ def _multi_tensor_adam(params: List[Tensor],
+        per_device_and_dtype_tensors[key][2].append(exp_avgs[i])
+        per_device_and_dtype_tensors[key][3].append(exp_avg_sqs[i])
+        if max_exp_avg_sqs:
+            per_device_and_dtype_tensors[key][4].append(exp_avg_sqs[i])
```
my bad, found a silly mistake thanks to the updated test_optim.py
```suggestion
            per_device_and_dtype_tensors[key][4].append(max_exp_avg_sqs[i])
```

crcrpar committed:
pass max_exp_avg_sq

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Squeeze `step` Tensor in test.

When either `fused` or `capturable` is `True`, `step` is 1D Tensor.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
check states as well as param and grad

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
stop using `torch.optim._fused`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
init foreach/fused optimizer using dunder init args

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Change dispatcher to float&fp16/bf16

because `torch._foreach_sqrt` uses `AT_DISPATCH_FLOATING_TYPES_AND2`.

Also, wrap `std` math functions into functors likewise foreach.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
just call `std:pow`, `std::max`, and `std::sqrt`

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Add `fused` to docstring

also, make sure param's dtype is floating point

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
deliberately copy kwargs

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remote vain context

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
import deepcopy to avoid copy being masked

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
remove `torch.optim._fused` namespace

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
cosmetic

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
write grad only when necessary

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
use lerp following Natalia's comment

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
Revert "use lerp following Natalia's comment"

This reverts commit cebc5f475d7d735e527f01da8ec4f3244182eb49.

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
From my side, this PR is ready for another round of review.
I don't know what to do for the failing jobs as they don't look directly related to this pull request to me.

ngimel commented:
@malfet do you know what could be causing sccache errors?
```
2022-08-10T16:29:38.7230215Z sccache: error: failed to execute compile
2022-08-10T16:29:38.7230673Z sccache: caused by: error reading compile response from server
2022-08-10T16:29:38.7232165Z sccache: caused by: Failed to read response header
2022-08-10T16:29:38.7232619Z sccache: caused by: failed to fill whole buffer
2022-08-10T16:29:38.7234971Z caffe2/CMakeFiles/torch_cuda_cu.dir/build.make:1357: recipe for target 'caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/FusedAdamKernel.cu.o' failed
2022-08-10T16:29:38.7235433Z make[2]: *** [caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/FusedAdamKernel.cu.o] Error 2
```

crcrpar commented:
linux-xenial-cuda11_3-py3_7-gcc7-deloy /build failed against for 1aff1ed: https://github.com/pytorch/pytorch/runs/7849107006?check_suite_focus=true#step:10:28829

```console
sccache: error: failed to execute compile
sccache: caused by: error reading compile response from server
sccache: caused by: Failed to read response header
sccache: caused by: failed to fill whole buffer
caffe2/CMakeFiles/torch_cuda_cu.dir/build.make:1266: recipe for target 'caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/FusedAdamKernel.cu.o' failed
make[2]: *** [caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/FusedAdamKernel.cu.o] Error 2
CMakeFiles/Makefile2:9360: recipe for target 'caffe2/CMakeFiles/torch_cuda_cu.dir/all' failed
make[1]: *** [caffe2/CMakeFiles/torch_cuda_cu.dir/all] Error 2
```
I think this is the same as or really similar to https://github.com/pytorch/pytorch/pull/81705#issuecomment-1211535683

malfet commented:
> @malfet do you know what could be causing sccache errors?
> 
> ```
> 2022-08-10T16:29:38.7230215Z sccache: error: failed to execute compile
> 2022-08-10T16:29:38.7230673Z sccache: caused by: error reading compile response from server
> 2022-08-10T16:29:38.7232165Z sccache: caused by: Failed to read response header
> 2022-08-10T16:29:38.7232619Z sccache: caused by: failed to fill whole buffer
> 2022-08-10T16:29:38.7234971Z caffe2/CMakeFiles/torch_cuda_cu.dir/build.make:1357: recipe for target 'caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/FusedAdamKernel.cu.o' failed
> 2022-08-10T16:29:38.7235433Z make[2]: *** [caffe2/CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/FusedAdamKernel.cu.o] Error 2
> ```

Sorry for the late reply. To be frank, it wasn't at all clear to me, what's causing the failure, but after recreating the [PR](https://github.com/pytorch/pytorch/pull/83548) and sshing into it it became super simple. nvcc from cuda-11.3 needs more than 20 min to finish compiling the FusedAdamKernel.cu, i.e.:
```
$ time /usr/local/cuda/bin/nvcc -x cu -c /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/FusedAdamKernel.cu -o CMakeFiles/torch_cuda_cu.dir/__/aten/src/ATen/native/cuda/FusedAdamKernel.cu.o -I/var/lib/jenkins/workspace/build/aten/src -I/var/lib/jenkins/workspace/aten/src -I/var/lib/jenkins/workspace/build -I/var/lib/jenkins/workspace -I/var/lib/jenkins/workspace/cmake/../third_party/benchmark/include -I/opt/llvm/include -I/var/lib/jenkins/workspace/cmake/../third_party/cudnn_frontend/include -I/var/lib/jenkins/workspace/third_party/onnx -I/var/lib/jenkins/workspace/build/third_party/onnx -I/var/lib/jenkins/workspace/third_party/foxi -I/var/lib/jenkins/workspace/build/third_party/foxi -I/var/lib/jenkins/workspace/build/include -I/var/lib/jenkins/workspace/aten/src/THC -I/var/lib/jenkins/workspace/aten/src/ATen/cuda -I/var/lib/jenkins/workspace/build/caffe2/aten/src -I/var/lib/jenkins/workspace/aten/../third_party/catch/single_include -I/var/lib/jenkins/workspace/aten/src/ATen/.. -I/var/lib/jenkins/workspace/c10/cuda/../.. -I/var/lib/jenkins/workspace/c10/.. -I/var/lib/jenkins/workspace/build/nccl/include -I/var/lib/jenkins/workspace/third_party/tensorpipe -I/var/lib/jenkins/workspace/build/third_party/tensorpipe -I/var/lib/jenkins/workspace/third_party/tensorpipe/third_party/libnop/include -I/var/lib/jenkins/workspace/torch/csrc/api -I/var/lib/jenkins/workspace/torch/csrc/api/include -isystem /var/lib/jenkins/workspace/build/third_party/gloo -isystem /var/lib/jenkins/workspace/cmake/../third_party/gloo -isystem /var/lib/jenkins/workspace/cmake/../third_party/googletest/googlemock/include -isystem /var/lib/jenkins/workspace/cmake/../third_party/googletest/googletest/include -isystem /var/lib/jenkins/workspace/third_party/protobuf/src -isystem /opt/conda/include -isystem /var/lib/jenkins/workspace/third_party/gemmlowp -isystem /var/lib/jenkins/workspace/third_party/neon2sse -isystem /var/lib/jenkins/workspace/third_party/XNNPACK/include -isystem /var/lib/jenkins/workspace/third_party/ittapi/include -isystem /var/lib/jenkins/workspace/cmake/../third_party/eigen -isystem /usr/lib/openmpi/include/openmpi/opal/mca/event/libevent2021/libevent -isystem /usr/lib/openmpi/include/openmpi/opal/mca/event/libevent2021/libevent/include -isystem /usr/lib/openmpi/include/openmpi -isystem /usr/lib/openmpi/include -isystem /usr/local/cuda/include -isystem /var/lib/jenkins/workspace/third_party/ideep/mkl-dnn/third_party/oneDNN/include -isystem /var/lib/jenkins/workspace/third_party/ideep/include -isystem /var/lib/jenkins/workspace/third_party/ideep/mkl-dnn/include --expt-relaxed-constexpr --expt-extended-lambda --expt-extended-lambda -Xcompiler -fPIC -Xcompiler -Werror,-Wall,-Wextra,-Wno-unused-parameter,-Wno-unused-function,-Wno-unused-result,-Wno-unused-local-typedefs,-Wno-missing-field-initializers,-Wno-write-strings,-Wno-unknown-pragmas,-Wno-type-limits,-Wno-array-bounds,-Wno-unknown-pragmas,-Wno-sign-compare,-Wno-strict-overflow,-Wno-strict-aliasing,-Wno-error=deprecated-declarations,-Wno-missing-braces,-Wno-maybe-uninitialized,-Wno-strict-overflow -Xcompiler -pthread -DAT_PER_OPERATOR_HEADERS -DBUILD_SPLIT_CUDA -DHAVE_MALLOC_USABLE_SIZE=1 -DHAVE_MMAP=1 -DHAVE_SHM_OPEN=1 -DHAVE_SHM_UNLINK=1 -DIDEEP_USE_MKL -DMINIZ_DISABLE_ZIP_READER_CRC32_CHECKS -DONNXIFI_ENABLE_EXT=1 -DONNX_ML=1 -DONNX_NAMESPACE=onnx_torch -DTORCH_CUDA_CU_BUILD_MAIN_LIB -DTORCH_ENABLE_LLVM -DUSE_C10D_GLOO -DUSE_C10D_MPI -DUSE_C10D_NCCL -DUSE_CUDA -DUSE_DISTRIBUTED -DUSE_EXPERIMENTAL_CUDNN_V8_API -DUSE_EXTERNAL_MZCRC -DUSE_RPC -DUSE_TENSORPIPE -D_FILE_OFFSET_BITS=64 -Dtorch_cuda_cu_EXPORTS -Xfatbin -compress-all -DONNX_NAMESPACE=onnx_torch -gencode arch=compute_52,code=sm_52 -Xcudafe --diag_suppress=cc_clobber_ignored,--diag_suppress=integer_sign_change,--diag_suppress=useless_using_declaration,--diag_suppress=set_but_not_used,--diag_suppress=field_without_dll_interface,--diag_suppress=base_class_has_different_dll_interface,--diag_suppress=dll_interface_conflict_none_assumed,--diag_suppress=dll_interface_conflict_dllexport_assumed,--diag_suppress=implicit_return_from_non_void_function,--diag_suppress=unsigned_compare_with_zero,--diag_suppress=declared_but_not_referenced,--diag_suppress=bad_friend_decl -Xfatbin -compress-all -Wno-deprecated-gpu-targets -DCUDA_HAS_FP16=1 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -O3 -DNDEBUG -DCAFFE2_USE_GLOO -DTH_HAVE_THREAD -DTORCH_CUDA_CU_BUILD_MAIN_LIB -std c++14 

???
```

And we set SCCACHE_IDLE_TIMEOUT to 20 min here
https://github.com/pytorch/pytorch/blob/31d4b6f52a9f74526f4f666348029da260254ea5/.jenkins/pytorch/common-build.sh#L19

One can update torch-deploy job to run against 11.6 (because why not), but still having files that needs more than 20 min to compile by 11.3 is slightly wrong, don't you think.

cc: @ZainRizvi @janeyx99 can we do something to improve UX here?

malfet commented:
Following PR will move deploy build to 11.6: https://github.com/pytorch/pytorch/pull/83572

crcrpar committed:
fix some wrongly rebased lines

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar committed:
avoid fused and differentiable being True simultaneously

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

crcrpar commented:
Thank you for taking a look @malfet 

> One can update torch-deploy job to run against 11.6 (because why not), but still having files that needs more than 20 min to compile by 11.3 is slightly wrong, don't you think.

According to nvcc folks, I don't have menas to speed up the build of this file from the implementation or whatever. So I'm kind of lost about what to do

ngimel commented:
Can we put amsgrad and non-amsgrad template instantitations in the different files (that'll hopefully be 2x reduction per file) https://github.com/pytorch/pytorch/pull/81705/files#diff-9a98433761fadd603e920775abbaa6fa195a464e9c16f5093d51de23e8337db7R214 ? 
And yes, @malfet what should be done to run torch deploy on cuda 11.6?

malfet commented:
> Can we put amsgrad and non-amsgrad template instantitations in the different files (that'll hopefully be 2x reduction per file) https://github.com/pytorch/pytorch/pull/81705/files#diff-9a98433761fadd603e920775abbaa6fa195a464e9c16f5093d51de23e8337db7R214 ?

Though can we please keep template implementation in .cuh file?

> And yes, @malfet what should be done to run torch deploy on cuda 11.6?

torchdeploy is running using 11.6 after https://github.com/pytorch/pytorch/pull/83572, but I have not looked at the compilation speed for this one

crcrpar committed:
Split compile into two files

finger-crossing so that CUDA 11.3 could build each file within 20
minutes...

Signed-off-by: Masaki Kozuki <mkozuki@nvidia.com>

stas00 commented:
Thank you for the heads up, @ngimel - happy you have been working on implementing this! Thank you!

crcrpar commented:
By splitting files, the previously failing CUDA 11.3 job looks fine but another CUDA 11.3 of [linux-xenial-cuda11.3-py3.7-gcc7-bazel-test / build-and-test](https://github.com/pytorch/pytorch/runs/7985450807?check_suite_focus=true#logs) seems failing.

[the raw log](https://pipelines.actions.githubusercontent.com/serviceHosts/7d146c05-69c3-4c20-a0e7-818111670117/_apis/pipelines/1/runs/2196050/signedlogcontent/18?urlExpires=2022-08-24T05%3A14%3A52.1486028Z&urlSigningMethod=HMACV1&urlSignature=hiUmfaptGgdNzoNDTxRxe4yMRiY3tZKDwfa5Yjf7IM4%3D) howeve doesn't seem informative

ngimel commented:
Looks like bazel is still on 11.3 (and failing).

crcrpar commented:
@pytorchmergebot rebase

crcrpar commented:
With https://github.com/pytorch/pytorch/pull/84314 and separate files for with and without amsgrad, CI looks green

ngimel commented:
Awesome!

ngimel commented:
@pytorchbot merge -f "rocm failure pre-existing"

ngimel commented:
@pytorchbot merge -f "rocm failure pre-existing"

dagitses commented:
@pytorchbot --help

dagitses commented:
@pytorchbot revert --message='broke internal builds, details to come' --classification=ghfirst

dagitses commented:
I know this isn't very detailed right now but what I'm seeing in the internal stack trace is the following failure: `AttributeError: 'Adam' object has no attribute 'fused'`, if that means anything to you.

crcrpar commented:
@dagitses I discuss this with @ngimel offline and I think I figured out the cause, just fyi

rejoicesyc commented:
Hey @crcrpar , thanks for your nice work, but I wonder why double type args are used like learning rate and weight decay, which made the kernel much slower since fp64 became a performance bottleneck.

crcrpar commented:
Following https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex
>  Floating point numbers are usually implemented using double in C
